src/transports/ws.rs

```diff
@@ -257,56 +257,94 @@ fn as_data_stream<T: Unpin + futures::AsyncRead + futures::AsyncWrite>(
     })
 }
 
+enum ParsedMessage {
+    /// Represents a JSON-RPC notification
+    Notification(rpc::Notification),
+    /// Represents a valid JSON-RPC response
+    Response(rpc::Response),
+    /// Represents an malformed JSON-RPC response, although containing the request `id`
+    InvalidResponse(rpc::Id),
+}
+
+fn parse_message(data: &[u8]) -> Option<ParsedMessage> {
+    if let Ok(notification) = helpers::to_notification_from_slice(data) {
+        Some(ParsedMessage::Notification(notification))
+    } else if let Ok(response) = helpers::to_response_from_slice(data) {
+        Some(ParsedMessage::Response(response))
+    } else if let Some(id) = helpers::response_id_from_slice(data) {
+        Some(ParsedMessage::InvalidResponse(id))
+    } else {
+        None
+    }
+}
```

Comment:
I guess it might be better to simply use untagged enum `ParsedMessage` with default serde deserialize behaviour? Given some tests are written I think it should be simpler to reason about and possibly also faster than manually attempting to parse the response couple of times.